flow/controllers/base_controller.py

```diff
@@ -235,4 +285,93 @@ def safe_velocity(self, env):
 
         v_safe = 2 * h / env.sim_step + dv - this_vel * (2 * self.delay)
 
+        # check for speed limit  FIXME: this is not called
+        # this_edge = env.k.vehicle.get_edge(self.veh_id)
+        # edge_speed_limit = env.k.network.speed_limit(this_edge)
+
+        if this_vel > v_safe:
+            print(
+                "=====================================\n"
+                "Speed of vehicle {} is greater than safe speed. Safe velocity "
+                "clipping applied.\n"
+                "=====================================".format(self.veh_id))
+
         return v_safe
+
+    def get_obey_speed_limit_action(self, env, action):
+        """Perform the "obey_speed_limit" failsafe action.
+
+        Checks if the computed acceleration would put us above edge speed limit.
+        If it would, output the acceleration that would put at the speed limit
+        velocity.
+
+        Parameters
+        ----------
+        env : flow.envs.Env
+            current environment, which contains information of the state of the
+            network at the current time step
+        action : float
+            requested acceleration action
+
+        Returns
+        -------
+        float
+            the requested action clipped by the speed limit
+        """
+        # check for speed limit
+        this_edge = env.k.vehicle.get_edge(self.veh_id)
+        edge_speed_limit = env.k.network.speed_limit(this_edge)
+
+        this_vel = env.k.vehicle.get_speed(self.veh_id)
+        sim_step = env.sim_step
+
+        if this_vel + action * sim_step > edge_speed_limit:
```

Comment:
Would it be useful to create a `env.k.vehicle.get_next_speed(accel)` function? Since this operation is done at several places

Reply:
I implemented it (here: [3161580](https://github.com/flow-project/flow/pull/981/commits/316158059eaa626726cfe7b5a4e01d8cb607b392)) but then it caused other problems for not having `this_vel` and `sim_step` in scope anymore, so I will keep it as is.